app/transactions/enrollments/create_transaction.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+module Enrollments
+  # create enrollments after prepare params
+  class CreateTransaction
+    include Dry::Transaction
+
+    step :validate
+    tee :prepare
+    step :create
+
+    def validate(input, mission:)
+      return Success(input) if mission.max_member_count.nil?
+      return Success(input) if mission.members.count < mission.max_member_count
+
+      failure_message = I18n.t('enrollments.create.max_member_count_reached')
+      Failure(failure_message)
+    end
+
+    def prepare(input, mission:)
+      return Success(input) unless mission.genre == 'regulated'
+
+      input['end_time'] = input['start_time'].last.to_datetime + 90.minutes
+      input['start_time'] = input['start_time'].first.to_datetime
```

Comment:
This transaction has knowledge it shouldn't have:
It knows that the end_time should be based on the last `input['start_time']`
It knows that the start_time should be based on the first `input['start_time']`
This knowledge creates an invisible dependency, that is invisible to the caller : the caller has no way to know that a specific array order is needed in order to properly do the transaction.

You should use dependency injection here, as giving an explicit argument to the transaction. In doing so, you won't need the `prepare` step anymore.

You would then create an Enrollement by choosing explicit items from the `input` hash

Reply:
If i inject an explicit argument, i prepare this argument in controller. The aim of transaction is reduce controller. 
If i sort start_time in transaction, is acceptable?

Reply:
Yeah that's a good intent. But the Controller 'controls' what's going on. Other objects are responsible for the actions they are asked to do, and depending and what they are asked to do, we may pass them different arguments. So it's better to give an object what it needs for a given action, rather that giving that object the knowledge of WHO is asking the action.
E.g.: here, the transaction knows that the asked action comes from a given action in a given controller. So this context could be different if that action was asked in another action from an another controller.

I think it's a good idea to see the interaction as a discussion between the controller and the transaction:
- Controller: Hey could you make me an enrollement record?
- Transaction: Yeah, on what `mission` ?
- Controller: On `this` one please
- Transaction: And on what time sould it start, and what time should it end?
- Controller: From 10am to 1pm
- Transaction: OK done

Sorting the given array would be better yes, but  the problem stays the same: you are asking the Transaction to search for a given record ; but you also asked it to create an enrollment.
So It's better if you give the Transaction what it needs to work explicitely. As a nice side effect, that would document the use of this transaction: someone reading the code in the controller would know what this transaction needs (e.g.: e given record, an array, or a hash...), without knowing how it works.

That can allow us, for example, to have the minimum amount of information needed to modify the controller as we want, without modifying the transaction's behaviour without knowing it (eg.: by changing some data structures that the transaction would not recognize)

Reply:
I guess it's actually the same logic as using `.with_step_args` : you explicitely pass the argument you want the transaction to work with

Reply:
done